tap_zendesk/client.py

```diff
@@ -95,56 +74,133 @@ def get_url_params(
+        """
+        Handle the Zendesk API only allowing 700 API calls per minute.
+        """
+        headers = response.headers
+        rate_limit_remain = headers.get('x-rate-limit-remaining', headers.get('ratelimit-remaining'))
+        if rate_limit_remain:
+            rate_limit_remain = int(rate_limit_remain)
+            rate_limit = int(headers.get('x-rate-limit', headers.get('ratelimit-limit')))
+            rate_limit_resets_in_s = None
+            self.logger.debug(f"Remaining rate limit: {rate_limit_remain}/{rate_limit}" +
+                             (f" (reset in {rate_limit_resets_in_s}s)" if rate_limit_resets_in_s is not None else ""))
+            if rate_limit_remain <= min_remain_rate_limit:
+                seconds_to_sleep = int(rate_limit_resets_in_s) if rate_limit_resets_in_s is not None else 60
+                self.logger.warning(f"API rate limit exceeded (rate limit: {rate_limit}, remain: {rate_limit_remain}, "
+                             f"min remain limit: {min_remain_rate_limit}). "
+                             f"Tap will retry the data collection after {seconds_to_sleep} seconds.")
+                sleep(seconds_to_sleep)
+        else:
+            raise Exception(f"rate-limit-remaining not found in response header: {headers}")
+
+    def _request(
         self,
-        row: dict,
-        context: dict | None = None,  # noqa: ARG002
-    ) -> dict | None:
-        """As needed, append or transform raw data to match expected structure.
+        prepared_request: PreparedRequest,
+        context: Optional[Dict[str, Any]] = None,
+    ) -> Response:
+        """Send a HTTP request with rate limiting.
 
         Args:
-            row: An individual record from the stream.
-            context: The stream context.
+            prepared_request: The prepared HTTP request.
+            context: Stream partition or context dictionary.
 
         Returns:
-            The updated record dictionary, or ``None`` to skip the record.
+            The HTTP response object.
         """
-        # TODO: Delete this method if not needed.
-        return row
+
+        response = self.requests_session.send(
+            prepared_request,
+            timeout=self.timeout,
+            allow_redirects=self.allow_redirects,
+        )
+
+        self.last_request_time = time.time()
+
+        self._write_request_duration_log(
+            endpoint=self.path,
+            response=response,
+            context=context,
+            extra_tags={"url": prepared_request.path_url}
+            if self._LOG_REQUEST_METRIC_URLS
+            else None,
+        )
+
+        # Check for 404 Not Found and log a warning
+        if response.status_code == 404:
+            self.logger.warning(f"Received 404 for URL: {response.url}")
+            return None
+
+        # Rate throttling check
+        self.check_rate_throttling(response, min_remain_rate_limit=10)
```

Comment:
Let's use the minimum rate limit as a config parameter and ignore the check if it's not set.

Reply:
Sure, done in [37ad58c](https://github.com/Automattic/tap-zendesk-singer-sdk/pull/1/commits/37ad58cdce2fcd4ed6dac627e7db5ba59e1b5aea)